homeassistant/components/mobile_app/webhook.py

```diff
@@ -0,0 +1,192 @@
+                                 CONF_WEBHOOK_ID, HTTP_BAD_REQUEST)
+from homeassistant.core import EventOrigin
+from homeassistant.exceptions import (HomeAssistantError, ServiceNotFound,
+                                      TemplateError)
+from homeassistant.helpers import template
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.helpers.storage import Store
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.loader import get_platform
+
+from .const import (ATTR_APP_COMPONENT, DATA_DELETED_IDS,
+                    ATTR_DEVICE_NAME, ATTR_EVENT_DATA, ATTR_EVENT_TYPE,
+                    DATA_REGISTRATIONS, ATTR_TEMPLATE, ATTR_TEMPLATE_VARIABLES,
+                    ATTR_WEBHOOK_DATA, ATTR_WEBHOOK_ENCRYPTED,
+                    ATTR_WEBHOOK_ENCRYPTED_DATA, ATTR_WEBHOOK_TYPE,
+                    CONF_CLOUDHOOK_ID, CONF_CLOUDHOOK_URL, CONF_SECRET,
+                    DOMAIN, HTTP_X_CLOUD_HOOK_ID, HTTP_X_CLOUD_HOOK_URL,
+                    WEBHOOK_PAYLOAD_SCHEMA, WEBHOOK_SCHEMAS,
+                    WEBHOOK_TYPES, WEBHOOK_TYPE_CALL_SERVICE,
+                    WEBHOOK_TYPE_FIRE_EVENT, WEBHOOK_TYPE_RENDER_TEMPLATE,
+                    WEBHOOK_TYPE_UPDATE_LOCATION,
+                    WEBHOOK_TYPE_UPDATE_REGISTRATION)
+
+from .helpers import (device_context, _decrypt_payload, empty_okay_response,
+                      safe_device, savable_state)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_device(hass: HomeAssistantType, store: Store, device: Dict) -> None:
+    """Register the webhook for a device and loads the app component."""
+    device_name = 'Mobile App: {}'.format(device[ATTR_DEVICE_NAME])
+    webhook_id = device[CONF_WEBHOOK_ID]
+    webhook_register(hass, DOMAIN, device_name, webhook_id,
+                     partial(handle_webhook, store))
+
+    if ATTR_APP_COMPONENT in device:
+        load_platform(hass, device[ATTR_APP_COMPONENT], DOMAIN, {},
+                      {DOMAIN: {}})
+
+
+async def handle_webhook(store: Store, hass: HomeAssistantType,
+                         webhook_id: str, request: Request) -> Response:
+    """Handle webhook callback."""
+    if webhook_id in hass.data[DOMAIN][DATA_DELETED_IDS]:
+        return Response(status=410)
+
+    device = hass.data[DOMAIN][DATA_REGISTRATIONS][webhook_id]
+
+    try:
+        req_data = await request.json()
+    except ValueError:
+        _LOGGER.warning('Received invalid JSON from mobile_app')
+        return json_response([], status=HTTP_BAD_REQUEST)
+
+    try:
+        req_data = WEBHOOK_PAYLOAD_SCHEMA(req_data)
+    except vol.Invalid as ex:
+        err = vol.humanize.humanize_error(req_data, ex)
+        _LOGGER.error('Received invalid webhook payload: %s', err)
+        return empty_okay_response()
+
+    headers = {}
+
+    via_cloud = is_cloudhook_request(request)
+
+    logged_in = async_is_logged_in(hass)
+
+    if via_cloud is False and logged_in is True:
```

Comment:
Can we drop this logic for now?

There are a few things that are not good:

 - We should not check `async_is_logged_in` but instead `active_subscription`, which I have in a PR.
 - We should not create a cloud url as a side-effect of another action. Instead we could add in the future a header: `cloud-available: true`, to indicate that they can send a `update_registration` command to get a cloudhook url.